### PR TITLE
CB-3268 feat: admin users pagination

### PR DESCRIPTION
**Pull request by Wroud**

### CODE REVIEW

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java [hunk 1]
```diff
@@ -291,19 +291,76 @@ public SMUser getCurrentUser() throws DBException {
...
+            }
+            if (!CommonUtils.isEmpty(userIdMask)) {
+                whereString += (whereString != "\nWHERE" ? "AND" : "") + " USER_ID LIKE ?";
+            }
+            if (enabledState != null) {
+                whereString += (whereString != "\nWHERE" ? "AND" : "") + " IS_ACTIVE=?";
+            }
+            try (PreparedStatement dbStat = dbCon.prepareStatement(database.normalizeTableNames(
+                    "SELECT COUNT(*) FROM {table_prefix}CB_USER" + (whereString != "\nWHERE" ? whereString : "")))) {
+                int parameterIndex = 1;
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: create separate buildUsersFilter method

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java [hunk 2]
```diff
@@ -291,19 +291,76 @@ public SMUser getCurrentUser() throws DBException {
...
+    }
+
     @NotNull
     @Override
-    public SMUser[] findUsers(String userNameMask) throws DBCException {
+    public SMUser[] findUsers(Integer first, String after, String userIdMask, Boolean enabledState)
+            throws DBCException {
         try (Connection dbCon = database.openConnection()) {
             Map<String, SMUser> result = new LinkedHashMap<>();
+            String whereString = "\nWHERE";
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: and use buildUsersFilter here too

Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: use StringBuilder

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java [hunk 3]
```diff
@@ -291,19 +291,76 @@ public SMUser getCurrentUser() throws DBException {
...
+            if (!CommonUtils.isEmpty(after)) {
+                whereString += " USER_ID > ?";
+            }
+            if (!CommonUtils.isEmpty(userIdMask)) {
+                whereString += (whereString != "\nWHERE" ? "AND" : "") + " USER_ID LIKE ?";
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: very strange ternary operator

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java [hunk 4]
```diff
@@ -291,19 +291,76 @@ public SMUser getCurrentUser() throws DBException {
...
+            }
+            String usersTableWhere = whereString;
+            if (enabledState != null) {
+                usersTableWhere += (usersTableWhere != "\nWHERE" ? "AND" : "") + " IS_ACTIVE=?";
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: same here

File: server/bundles/io.cloudbeaver.model/src/io/cloudbeaver/model/pagination/GQLConnection.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * limitations under the License.
+ */
+package io.cloudbeaver.model.pagination;
+
+import java.util.List;
+
+/**
+ * GraphQL connection
+ */
+public class GQLConnection<T> {
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.model/src/io/cloudbeaver/model/pagination/GQLConnection.java]: naming is not clear. may be rename to GQLPage?

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java [hunk 6]
```diff
@@ -291,19 +291,76 @@ public SMUser getCurrentUser() throws DBException {
...
     @NotNull
     @Override
-    public SMUser[] findUsers(String userNameMask) throws DBCException {
+    public SMUser[] findUsers(Integer first, String after, String userIdMask, Boolean enabledState)
+            throws DBCException {
         try (Connection dbCon = database.openConnection()) {
             Map<String, SMUser> result = new LinkedHashMap<>();
+            String whereString = "\nWHERE";
+            if (!CommonUtils.isEmpty(after)) {
+                whereString += " USER_ID > ?";
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: in my opinion it is strange that we want to compare strings in this way

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java [hunk 7]
```diff
@@ -291,19 +291,76 @@ public SMUser getCurrentUser() throws DBException {
         return getUserById(getUserIdOrThrow());
     }
 
+    @Override
+    public int countUsers(String after, String userIdMask, Boolean enabledState) throws DBCException {
+        try (Connection dbCon = database.openConnection()) {
+            String whereString = "\nWHERE";
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: filter construction logic seems overcomplicated and needs to be simplified

File: server/bundles/io.cloudbeaver.service.admin/src/io/cloudbeaver/service/admin/impl/WebServiceAdmin.java [hunk 8]
```diff
@@ -66,16 +69,38 @@ public class WebServiceAdmin implements DBWServiceAdmin {
...
-            for (SMUser smUser : webSession.getAdminSecurityController().findUsers(userName)) {
-                webUsers.add(new AdminUserInfo(webSession, new WebUser(smUser)));
+            List<GQLConnectionEdge<AdminUserInfo>> edges = new ArrayList<>();
+            for (SMUser smUser : webSession.getAdminSecurityController().findUsers(first, after, userIdMask, enabledState)) {
+                edges.add(
+                        new GQLConnectionEdge<>(new AdminUserInfo(webSession, new WebUser(smUser)),
+                                smUser.getUserId()));
             }
-            return webUsers;
+            int totalCount = webSession.getAdminSecurityController().countUsers(null, userIdMask, enabledState);
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.admin/src/io/cloudbeaver/service/admin/impl/WebServiceAdmin.java]: we can return search results with already known count, and we don't need to additionally call the count method

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java [hunk 9]
```diff
@@ -291,19 +291,76 @@ public SMUser getCurrentUser() throws DBException {
...
+                if (!CommonUtils.isEmpty(after)) {
+                    dbStat.setString(parameterIndex++, after);
+                }
+                if (!CommonUtils.isEmpty(userIdMask)) {
+                    dbStat.setString(parameterIndex++, "%" + userIdMask + "%");
                 }
+                if (enabledState != null) {
+                    dbStat.setString(parameterIndex++, enabledState ? CHAR_BOOL_TRUE : CHAR_BOOL_FALSE);
+                }
+                dbStat.setInt(parameterIndex++, first);
```
Comment (alexander-skoblikov) [inline server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: why first used as limit?

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java (hunk 1 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: buildUsersFilter created

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java (hunk 2 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: removed

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java (hunk 3 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: removed

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java (hunk 4 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: removed

File: server/bundles/io.cloudbeaver.model/src/io/cloudbeaver/model/pagination/GQLConnection.java (hunk 5 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.model/src/io/cloudbeaver/model/pagination/GQLConnection.java]: removed

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java (hunk 6 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: replaced with limit/offset pagination

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java (hunk 7 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: moved to separate function

File: server/bundles/io.cloudbeaver.service.admin/src/io/cloudbeaver/service/admin/impl/WebServiceAdmin.java (hunk 8 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.admin/src/io/cloudbeaver/service/admin/impl/WebServiceAdmin.java]: the count is not required as a part of the slice function

File: server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java (hunk 9 above)
Comment (Wroud) [reply server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/CBEmbeddedSecurityController.java]: removed

**Review by devnaumov — APPROVED** [review]

**Pull request merged**